File: onedal/data_management/csr_table/csr_table.cpp
/*******************************************************************************
* Copyright 2023 Intel Corporation
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*     http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*******************************************************************************/

#include <memory>

#include <pybind11/pybind11.h>

#include "onedal/common.hpp"
#include "onedal/common/dtype_dispatcher.hpp"
#include "onedal/data_management/table/table_iface.hpp"

#include "oneapi/dal/array.hpp"
#include "oneapi/dal/common.hpp"

#include "oneapi/dal/table/csr.hpp"
#include "oneapi/dal/table/common.hpp"

#include "oneapi/dal/detail/common.hpp"

namespace py = pybind11;

namespace oneapi::dal::python::data_management {

inline void instantiate_csr_indexing(py::module& pm) {
    constexpr const char name[] = "sparse_indexing";
    py::enum_<dal::sparse_indexing> py_indexing(pm, name);
    py_indexing.value("one_based", dal::sparse_indexing::one_based);
    py_indexing.value("zero_based", dal::sparse_indexing::zero_based);
    py_indexing.export_values();
}

template <typename Type, typename Table = dal::csr_table>
inline void instantiate_csr_constructor_impl(py::class_<Table>& py_table) {
    py_table.def(py::init([](const dal::array<Type>& non_zeros,
                             const dal::array<std::int64_t>& col_indices,
                             const dal::array<std::int64_t>& row_offsets,
                             std::int64_t col_count) {
        return csr_table::wrap<Type>(non_zeros, col_indices, row_offsets, col_count);
    }));
    py_table.def(py::init([](const dal::array<Type>& non_zeros,
                             const dal::array<std::int64_t>& col_indices,
                             const dal::array<std::int64_t>& row_offsets,
                             std::int64_t col_count,
                             dal::sparse_indexing sp) {
        return csr_table::wrap<Type>(non_zeros, col_indices, row_offsets, col_count, sp);
    }));
}

template <typename Table, typename... Types>
inline void instantiate_csr_constructor(py::class_<Table>& py_table,
                                        const std::tuple<Types...>* const = nullptr) {
    static_assert(std::is_same_v<Table, csr_table>);
    return detail::apply(
        [&](auto type_tag) -> void {
            using type_t = std::decay_t<decltype(type_tag)>;
            instantiate_csr_constructor_impl<type_t>(py_table);
        },
        Types{}...);
}

detail::csr_table_iface* get_csr_table_iface_impl(detail::table_iface* table) {
    return dynamic_cast<detail::csr_table_iface*>(table);
}

template <typename Table>
inline std::shared_ptr<detail::csr_table_iface> get_interface(const Table& table) {
    auto pimpl = detail::pimpl_accessor{}.get_pimpl(table);
    auto csr_iface_ptr = get_csr_table_iface_impl(pimpl.get());
    return std::shared_ptr<detail::csr_table_iface>{ pimpl, csr_iface_ptr };
}

template <typename Type>
dal::array<Type> get_data(const dal::csr_table& table) {
    const std::int64_t elem_size = //
        detail::integral_cast<std::int64_t>(sizeof(Type));

    auto impl = get_interface(table);
    const auto count = table.get_non_zero_count();

    dal::array<dal::byte_t> data = impl->get_data();

    const std::int64_t size = //
        detail::check_mul_overflow(count, elem_size);

    if (size != data.get_count()) {
        throw std::length_error("Incorrect data size");
    }

    if (data.has_mutable_data()) {
        dal::byte_t* raw_ptr = data.get_mutable_data();
        Type* ptr = reinterpret_cast<Type*>(raw_ptr);
        return dal::array<Type>(data, ptr, count);
    }
    else {
        const dal::byte_t* raw_ptr = data.get_data();
        const Type* ptr = reinterpret_cast<const Type*>(raw_ptr);
        return dal::array<Type>(data, ptr, count);
    }
}

py::object get_data_array(const dal::csr_table& table) {
    const table_metadata& meta = table.get_metadata();
    const data_type dtype = meta.get_data_type(0l);

    return detail::dispatch_by_data_type(dtype, [&](auto type_tag) -> py::object {
        using type_t = std::decay_t<decltype(type_tag)>;
        auto array = get_data<type_t>(table);
        return py::cast(std::move(array));
    });
}

void instantiate_csr_table(py::module& pm) {
    instantiate_csr_indexing(pm);

    constexpr const char name[] = "csr_table";
    py::class_<dal::csr_table> py_csr_table(pm, name);

    //py_csr_table.def(py::init<dal::table>());

    // Need to be fixed. No such symbol in the library
    py_csr_table.def(py::init([](const dal::table& t) {
        if (t.get_kind() != csr_table::kind()) {
            throw std::runtime_error("Unable to cast to csr table");
        }
        const auto& casted = reinterpret_cast<const dal::csr_table&>(t);
        return dal::csr_table(casted);
    }));

    py_csr_table.def("get_indexing", &dal::csr_table::get_indexing);
    py_csr_table.def("get_non_zero_count", [](const dal::csr_table& table) -> std::int64_t {
        return table.get_non_zero_count();
    });

    py_csr_table.def("get_data", [](const dal::csr_table& table) -> py::object {
        return get_data_array(table);
    });

    py_csr_table.def("get_row_offsets", [](const dal::csr_table& table) -> py::object {
        auto iface = get_interface(table);
        auto array = iface->get_row_offsets();
        return py::cast(std::move(array));
    });

    py_csr_table.def("get_column_indices", [](const dal::csr_table& table) -> py::object {
        auto iface = get_interface(table);
        auto array = iface->get_column_indices();
        return py::cast(std::move(array));
    });

    instantiate_table_iface(py_csr_table);

    constexpr const supported_types_t* types = nullptr;
    instantiate_csr_constructor(py_csr_table, types);
}

} // namespace oneapi::dal::python::data_management

File: onedal/interop/csr_table.py
# ==============================================================================
# Copyright 2023 Intel Corporation
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ==============================================================================

import numpy as np
from scipy.sparse import csr_matrix, isspmatrix_csr

import onedal

from .array import from_array, to_array, to_common_policy, to_typed_array

table = onedal._backend.data_management.table
csr_table = onedal._backend.data_management.csr_table
csr_kind = onedal._backend.data_management.table_kind.csr
sparse_indexing = onedal._backend.data_management.sparse_indexing


def is_native_csr(entity) -> bool:
    if isinstance(entity, table):
        kind = entity.get_kind()
        return kind == csr_kind
    return isinstance(entity, csr_table)


def is_csr_entity(entity) -> bool:
    is_native = is_native_csr(entity)
    is_scipy = isspmatrix_csr(entity)
    return is_native or is_scipy


def assert_table(entity, matrix):
    assert is_native_csr(entity)
    assert isspmatrix_csr(matrix)
    row_count, col_count = matrix.shape
    assert entity.get_column_count() == col_count
    assert entity.get_row_count() == row_count


# Passing throw native entity
def to_csr_table_native(entity) -> csr_table:
    assert is_native_csr(entity)
    return csr_table(entity)


# Converting python entity to table
# TODO #1: Implement smarter logic for
# type conversion & device support
# TODO #2: Implement with `zero_based`
# indexing, it is a workaround for now.
# The glue logic between oneDAL & DAAL is
# suspicious for having bug with `zero_based`
def to_csr_table_python(entity) -> csr_table:
    assert isspmatrix_csr(entity)
    row_count, col_count = entity.shape

    def to_indices(arr, ids=[np.int64]):
        return to_typed_array(arr, ids)

    ids = to_indices(entity.indices + 1)
    ofs = to_indices(entity.indptr + 1)
    nz = to_array(entity.data)

    one_based = sparse_indexing.one_based
    ids, ofs, nz = to_common_policy(ids, ofs, nz)
    result = csr_table(nz, ids, ofs, col_count, one_based)
    assert col_count == result.get_column_count()
    assert row_count == result.get_row_count()
    assert_table(result, entity)
    return result


def to_csr_table(entity) -> csr_table:
    assert is_csr_entity(entity)
    if is_native_csr(entity):
        result = to_csr_table_native(entity)
    elif isspmatrix_csr(entity):
        result = to_csr_table_python(entity)
    else:
        raise ValueError("Not able to convert")
    assert is_native_csr(result)
    return result


# Converting onedal entity to python
def from_csr_table_native(entity) -> csr_matrix:
    assert is_native_csr(entity)
    entity = to_csr_table_native(entity)
    col_count = entity.get_column_count()
    row_count = entity.get_row_count()
    shape = (row_count, col_count)
    ids = from_array(entity.get_column_indices())
    ofs = from_array(entity.get_row_offsets())
    nz = from_array(entity.get_data())
    indexing = entity.get_indexing()
    zb = sparse_indexing.zero_based
    ob = sparse_indexing.one_based

    if indexing == ob:
        ids, ofs = ids - 1, ofs - 1
    else:
        assert indexing == zb

    data = (nz, ids, ofs)
    result = csr_matrix(data, shape)
    assert_table(entity, result)
    return result


# Passing matrix that is already in correct format
def from_csr_table_python(entity) -> csr_matrix:
    assert isspmatrix_csr(entity)
    return entity


def from_csr_table(table) -> csr_matrix:
    if is_native_csr(table):
        csr = csr_table(table)
        result = from_csr_table_native(csr)
    elif isspmatrix_csr(table):
        result = from_csr_table_python(table)
    else:
        raise ValueError("Not able to convert")
    assert isspmatrix_csr(result)
    return result
